tools/oeedger8r/Emitter.ml

```diff
@@ -967,274 +472,861 @@ let warn_size_and_count_params (fd : func_decl) =
+      ; ""
+      ; "#include <openenclave/bits/result.h>"
+      ; ""
+      ; "/**** User includes. ****/"
+      ; String.concat "\n" (oe_gen_user_includes ec.include_list)
+      ; ""
+      ; "/**** User defined types in EDL. ****/"
+      ; String.concat "\n" (oe_gen_user_types ec.comp_defs)
+      ; "/**** ECALL marshalling structs. ****/"
+      ; String.concat "\n" (oe_gen_ecall_marshal_structs ec.tfunc_decls)
+      ; "/**** OCALL marshalling structs. ****/"
+      ; String.concat "\n" (oe_gen_ocall_marshal_structs ec.ufunc_decls)
+      ; "/**** Trusted function IDs ****/"
+      ; String.concat "\n"
+          (emit_trusted_function_ids ec.tfunc_decls ec.enclave_name)
+      ; ""
+      ; "/**** Untrusted function IDs. ****/"
+      ; String.concat "\n"
+          (emit_untrusted_function_ids ec.ufunc_decls ec.enclave_name)
+      ; ""
+      ; sprintf "#endif // %s" guard_macro
+      ; "" ]
+    in
+    let fname = sprintf "%s_args.h" ec.file_shortnm in
+    let os = open_file fname dir in
+    fprintf os "%s" (String.concat "\n" content) ;
+    close_out os
   in
-  let oe_gen_ufunc_wrapper_prototypes (ufs : untrusted_func list) =
-    if ufs <> [] then
-      List.map
-        (fun f -> sprintf "%s;" (oe_gen_wrapper_prototype f.uf_fdecl false))
-        ufs
-    else ["/* There were no ocalls. */"]
+  (* We need to check Ptrs for Foreign or Struct types, then check
+     those against the user's structs, and then check if any members
+     should be deep copied. *)
+  let get_deepcopy a =
+    (* This transforms the [Struct] types embedded in a [composite_type
+       list] into an association list of the struct name to its parameter
+       list, which is much easier to deal with. *)
+    let get_structs =
+      filter_map (function
+        | StructDef s -> Some (s.sname, s.smlist)
+        | _ -> None )
+    in
+    let structs_list = get_structs ec.comp_defs in
+    let get_struct = function
+      | Ptr a -> (
+        match a with
+        | Foreign n | Struct n -> List.assoc_opt n structs_list
+        | _ -> None )
+      | _ -> None
+    in
+    let s = get_struct a in
+    let should_deepcopy_member (ptype, _) =
+      match ptype with
+      | PTPtr (_, attr) -> attr.pa_size <> empty_ptr_size
+      | PTVal _ -> false
+    in
+    if ep.experimental then
+      match s with
+      | Some s -> List.filter should_deepcopy_member s
+      | None -> []
+    else []
   in
-  let guard = sprintf "EDGER8R_%s_T_H" (String.uppercase ec.file_shortnm) in
-  let content =
-    [ sprintf "#ifndef %s" guard
-    ; sprintf "#define %s" guard
-    ; ""
-    ; "#include <openenclave/enclave.h>"
-    ; ""
-    ; sprintf "#include \"%s_args.h\"" ec.file_shortnm
-    ; ""
-    ; "OE_EXTERNC_BEGIN"
+  (* Prepare [input_buffer]. *)
+  let oe_prepare_input_buffer (fd : func_decl) (alloc_func : string) =
+    let oe_compute_buffer_size (buffer : string)
+        (predicate : parameter_type * declarator -> bool) (plist : pdecl list)
```

Comment:
This function signature is a good example of why I've soured on the explicit typing of parameters. I think we should instead just use deduced typing (coupled with way more comments than I have currently).

Reply:
There are couple of options here. You could use a type variable in the declaration for verbose types and have the compiler figure it out.

`let oe_compute_buffer_size (buffer:string) (predicate : 'a_predicate) (plist : pdecl list)`

Alternatively you could a type alias.

Reply:
I think i'm going to drop most argument types and instead rely on type deductions. After writing a lot of this code, I found that they add very little value (maybe none entirely).